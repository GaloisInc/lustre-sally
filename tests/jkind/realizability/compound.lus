-- Unrealizable
-- Requires proper flattening of complex types

type point = struct { x : int; y : int };

node main(p : point; z : int) returns (ok : bool);
var
  counter : int;
let
  counter = 0 -> 1 + pre counter;
  assert counter < 5 => p.x < p.y - 1;
  
  ok = p.x < z and z < p.y;
  --%PROPERTY ok;
  --%REALIZABLE p;
tel;
