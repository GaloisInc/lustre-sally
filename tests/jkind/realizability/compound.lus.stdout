Sally Model
===========

(define-state-type
   S
   (
     (Implies Bool)
     (counter Int)
     (ok Bool)
     (z Int)
     (|counter:init| Bool)
     (|gal-initializing| Bool)
     (|p.x| Int)
     (|p.y| Int)
     (|pre_counter| Int))
   (
     (z Int)
     (|p.x| Int)
     (|p.y| Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |counter:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|p.x| input.|p.x|)
      (= next.|p.y| input.|p.y|)
      (= next.z input.z)
      next.Implies
      (=
         next.ok
         (and
            (< next.|p.x| next.z)
            (< next.z next.|p.y|)))
      (= next.|pre_counter| state.counter)
      (ite
         state.|counter:init|
         (= next.counter 0)
         (=
            next.counter
            (+ 1 next.|pre_counter|)))
      (= next.|counter:init| false)
      (=
         next.Implies
         (=>
            (< next.counter 5)
            (<
               next.|p.x|
               (- next.|p.y| 1))))))

(query
   TS
   (or |gal-initializing| ok))

[Lustre] Validating properties:
[Lustre] Property ok... [Invalid] 
[Trace] 
ok:
 Step | p              | z | |-> | ok    
------+----------------+---+-----+-------
 1    | {x = 0, y = 2} | 0 |     | False 

[Lustre] Model status: [Invalid] 
